chain-signatures/node/src/protocol/sync/mod.rs

```diff
@@ -0,0 +1,630 @@
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use cait_sith::protocol::Participant;
+use rand::rngs::StdRng;
+use rand::seq::IteratorRandom;
+use rand::SeedableRng;
+use serde::{Deserialize, Serialize};
+use tokio::sync::{mpsc, oneshot, RwLock};
+use tokio::task::{JoinHandle, JoinSet};
+
+use crate::mesh::MeshState;
+use crate::node_client::NodeClient;
+use crate::rpc::NodeStateWatcher;
+use crate::storage::{PresignatureStorage, TripleStorage};
+
+use super::contract::primitives::{intersect, intersect_vec, Participants};
+use super::presignature::{Presignature, PresignatureId};
+use super::triple::{Triple, TripleId};
+
+/// How long for a synchronous [`SyncUpdate`] to get back a [`SyncView`] request before timeout.
+const REQUEST_UPDATE_TIMEOUT: Duration = Duration::from_millis(50);
+/// How long for a synchronous [`ProtocolRequest`] to get back a [`ProtocolResponse`] before timeout.
+const REQUEST_PROTOCOL_TIMEOUT: Duration = Duration::from_millis(50);
+/// The maximum number of protocol requests that can be queued. This will likely never get to
+/// this upper bound just purely based on our ProtocolConfig.
+const MAX_PROTOCOL_REQUESTS: usize = 128 * 1024;
+/// The maximum number of update requests that can be queued. This is pretty much just
+/// based on the number of participants in the network. If we have 1024 participants then
+/// our issue will more than likely not be the channel size.
+const MAX_SYNC_UPDATE_REQUESTS: usize = 1024;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SyncUpdate {
+    triples: Vec<TripleId>,
+    presignatures: Vec<PresignatureId>,
+}
+
+impl SyncUpdate {
+    pub fn empty() -> Self {
+        Self {
+            triples: Vec::new(),
+            presignatures: Vec::new(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.triples.is_empty() && self.presignatures.is_empty()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SyncView {
+    triples: HashSet<TripleId>,
+    presignatures: HashSet<PresignatureId>,
+}
+
+impl SyncView {
+    fn empty() -> Self {
+        Self {
+            triples: HashSet::new(),
+            presignatures: HashSet::new(),
+        }
+    }
+}
+
+struct SyncCache {
```

Comment:
Why do we need to cache data here? It adds complexity and introduces an additional source of truth for triples and presignatures. I would say we should always work with Redis directly since it is already a cashing solution.

Reply:
Is is a tmp solution until we extend data in Redis?